c10/core/SymIntArrayRef.h

```diff
@@ -13,6 +13,337 @@
 namespace c10 {
 using SymIntArrayRef = ArrayRef<SymInt>;
 
+template <>
+class ArrayRef<SymInt> final {
+  using T = SymInt;
+
+ public:
+  using iterator = const T*;
+  using const_iterator = const T*;
+  using size_type = size_t;
+  using value_type = T;
+
+  using reverse_iterator = std::reverse_iterator<iterator>;
+
+  enum Unchecked { UNCHECKED };
+
+ private:
+  /// The start of the array, in an external buffer.
+  const T* Data;
+
+  /// The number of elements.
+  size_type Length;
+
+  /// A bool saying if any element is symbolic
+  bool AnySymbolic;
+
+  void debugCheckNullptrInvariant() {
+    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(
+        Data != nullptr || Length == 0,
+        "created ArrayRef with nullptr and non-zero length! c10::optional relies on this being illegal");
+  }
+
+  void debugCheckAnySymbolicInvariant() {
+    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(
+        AnySymbolic == computeSymbolic(),
+        "created SymIntArrayRef with incorrect AnySymbolic tag; real tag is ",
+        !AnySymbolic);
+  }
+
+  bool computeSymbolic() const {
+    for (const auto& s : *this) {
+      if (s.is_symbolic())
+        return true;
+    }
+    return false;
+  }
+
+ public:
+  /// @name Constructors
+  /// @{
+
+  /// Construct an empty ArrayRef.
+  /* implicit */ constexpr ArrayRef()
+      : Data(nullptr), Length(0), AnySymbolic(false) {}
+
+  /// Construct an ArrayRef from a single element.
+  ArrayRef(const T& OneElt)
+      : Data(&OneElt), Length(1), AnySymbolic(OneElt.is_symbolic()) {}
+
+  /// Construct an ArrayRef from a pointer and length.
+  ArrayRef(const T* data, size_t length)
+      : Data(data), Length(length), AnySymbolic(computeSymbolic()) {
+    debugCheckNullptrInvariant();
+  }
+
+  ArrayRef(Unchecked, const int64_t* data, size_t length)
+      : Data(reinterpret_cast<const c10::SymInt*>(data)),
+        Length(length),
+        AnySymbolic(false) {
```

Comment:
Should we have an unchecked version for symbolic=true as well? Or we just don't need that today?